drivers/counter/timer_dtmr_cmsdk_apb.c

```diff
@@ -115,6 +112,33 @@ static int timer_dtmr_cmsdk_apb_set_alarm(struct device *dev,
 	return 0;
 }
 
+static u32_t timer_dtmr_cmsdk_apb_get_top_value(struct device *dev)
+{
+	struct timer_dtmr_cmsdk_apb_dev_data *data = dev->driver_data;
+
+	u32_t ticks = data->load;
+
+	return ticks;
+}
+
+static int timer_dtmr_cmsdk_apb_set_alarm(struct device *dev, u8_t chan_id,
```

Comment:
Driver should provide info structure on driver initialization (https://github.com/zephyrproject-rtos/zephyr/blob/4d8f735fc635d51491d71a489c65c3b94d67a209/include/counter.h#L84). Info structure contains information about number of channels. If channels = 0 then `set_channel_alarm` and `disable_channel_alarm` might not be provided.

Reply:
can you explain further what should be done here?

Reply:
you can see for example here: https://github.com/NXPmicro/zephyr/blob/d880c8a83c4a78d1be099635925249eddd20353a/drivers/counter/counter_mcux_rtc.c#L253

that there is a info structure that should be created with driver instance. Then this struct is used by coutner API for geting for example number of channels. `counter_set_channel_alarm` is using this function to check if given instance supports setting set_channel_alarm. If instance has no channels then it does not need to implement `set_channel` and `disable_channel`

Reply:
Gotcha, how is a channel defined?

Reply:
what do you mean?

Reply:
@nordic-krch Trying to understand what the definition of an channel is.

Reply:
channel usually maps to compare channel seen in timer peripheral. It is able to generate interrupt on given tick within counter period (period is being set by `set_top_value`). In pure hardware driver number of channels is usually limited by number of compare channels in the peripheral. In theory one could build multiple channels on single physical compare channel.

Reply:
And how is an alarm defined?

Reply:
currently this function is called in api `set_channel_alarm` so alarm is set for given channel. It's been renamed from `counter_set_alarm` because there is already deprecated function named `counter_set_alarm`.

Reply:
Is there anything that distinguishes a basic counter from an alarm?

On the CMSDK dual-timer hardware, trying to decide if I should treat the 2nd timer as just another counter or as an alarm.

Reply:
i would treat those as two counters. There is a dependency between running counter and alarm. e.g. alarm can be relative to current counter value or absolute. So if you have 2 timers which are independent and they do not have compare channels to be those are 2 counters.